Isosurfacing_3/include/CGAL/Isosurfacing_3/internal/Dual_contouring_internal.h

```diff
@@ -0,0 +1,362 @@
+
+        // compute edge intersections
+        std::vector<Point> edge_intersections;
+        std::vector<Vector> edge_intersection_normals;
+
+        for (int i = 0; i < Tables::N_EDGES; ++i) {
+            const auto& v0 = Tables::edge_to_vertex[i][0];
+            const auto& v1 = Tables::edge_to_vertex[i][1];
+
+            if (b[v0] != b[v1]) {  // e0
+                const FT u = (s[v0] - iso_value) / (s[v0] - s[v1]);
+                const Point p_lerp = CGAL::ORIGIN + ((1 - u) * pos[v0] + u * pos[v1]);
+                edge_intersections.push_back(p_lerp);
+                // const Vector n_lerp = (1 - u) * normals[v0] + u * normals[v1];
+                edge_intersection_normals.push_back(domain.gradient(p_lerp));
+            }
+        }
+
+        // MC Polygon Center of Mass
+        if (false) {
+            Vector com_vec(0, 0, 0);
+
+            for (int i = 0; i < edge_intersections.size(); ++i) {
+                com_vec += edge_intersections[i] - CGAL::ORIGIN;
+            }
+
+            Point p = CGAL::ORIGIN + com_vec / edge_intersections.size();
+            point = p;
+        }
+
+        // SVD QEM
+        if (true) {
+            Eigen::Matrix3d A;
+            A.setZero();
+            Eigen::Vector3d b;
+            b.setZero();
+            for (std::size_t i = 0; i < edge_intersections.size(); ++i) {
+                Eigen::Vector3d n_k = {edge_intersection_normals[i].x(), edge_intersection_normals[i].y(),
+                                       edge_intersection_normals[i].z()};
+                Eigen::Vector3d p_k = {edge_intersections[i].x(), edge_intersections[i].y(), edge_intersections[i].z()};
+                double d_k = n_k.transpose() * p_k;
+
+                Eigen::Matrix3d A_k = n_k * n_k.transpose();
+                Eigen::Vector3d b_k = d_k * n_k;
+                A += A_k;
+                b += b_k;
+            }
+
+            Eigen::JacobiSVD<Eigen::MatrixXd> svd(A, Eigen::ComputeThinU | Eigen::ComputeThinV);
+            // set threshold as in Peter Lindstrom's paper, "Out-of-Core
+            // Simplification of Large Polygonal Models"
+            svd.setThreshold(1e-3);
+
+            // Init x hat
+            Eigen::Vector3d x_hat;
+            x_hat << point.x(), point.y(), point.z();
+
+            // Lindstrom formula for QEM new position for singular matrices
+            Eigen::VectorXd v_svd = x_hat + svd.solve(b - A * x_hat);
+
+            point = Point(v_svd[0], v_svd[1], v_svd[2]);
+        }
+
+        // bbox
+        if (use_bbox) {
+            CGAL::Bbox_3 bbox = (CGAL::ORIGIN + pos[0]).bbox() + (CGAL::ORIGIN + pos[7]).bbox();
+
+            FT x = std::min<FT>(std::max<FT>(point.x(), bbox.xmin()), bbox.xmax());
+            FT y = std::min<FT>(std::max<FT>(point.y(), bbox.ymin()), bbox.ymax());
+            FT z = std::min<FT>(std::max<FT>(point.z(), bbox.zmin()), bbox.zmax());
+            point = Point(x, y, z);
```

Comment:
In principle you should use something like `geom_traits.construct_point_3_object()(x,y,z)`